core/src/main/java/nu/yona/server/subscriptions/service/UserAddService.java

```diff
@@ -0,0 +1,275 @@
+ *******************************************************************************/
+package nu.yona.server.subscriptions.service;
+
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.IntFunction;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import javax.annotation.PostConstruct;
+import javax.transaction.Transactional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import nu.yona.server.crypto.seckey.CryptoSession;
+import nu.yona.server.device.service.DeviceService;
+import nu.yona.server.device.service.UserDeviceDto;
+import nu.yona.server.exceptions.MobileNumberConfirmationException;
+import nu.yona.server.exceptions.UserOverwriteConfirmationException;
+import nu.yona.server.exceptions.YonaException;
+import nu.yona.server.goals.entities.BudgetGoal;
+import nu.yona.server.goals.entities.Goal;
+import nu.yona.server.goals.service.ActivityCategoryDto;
+import nu.yona.server.goals.service.ActivityCategoryService;
+import nu.yona.server.goals.service.GoalDto;
+import nu.yona.server.messaging.entities.MessageSource;
+import nu.yona.server.properties.YonaProperties;
+import nu.yona.server.sms.SmsTemplate;
+import nu.yona.server.subscriptions.entities.ConfirmationCode;
+import nu.yona.server.subscriptions.entities.User;
+import nu.yona.server.subscriptions.entities.UserAnonymized;
+import nu.yona.server.subscriptions.entities.UserPrivate;
+import nu.yona.server.subscriptions.entities.UserRepository;
+import nu.yona.server.util.Require;
+import nu.yona.server.util.TimeUtil;
+
+@Service
+public class UserAddService
+{
+	private enum UserSignUp
+	{
+		INVITED, FREE
+	}
+
+	private static final Logger logger = LoggerFactory.getLogger(UserAddService.class);
+
+	@Autowired(required = false)
+	private UserRepository userRepository;
+
+	@Autowired
+	private YonaProperties yonaProperties;
+
+	@Autowired(required = false)
+	private DeviceService deviceService;
+
+	@Autowired(required = false)
+	private ActivityCategoryService activityCategoryService;
+
+	@Autowired(required = false)
+	private WhiteListedNumberService whiteListedNumberService;
+
+	@Autowired(required = false)
+	private UserLookupService userLookupService;
+
+	@Autowired(required = false)
+	private UserUpdateService userUpdateService;
+
+	@PostConstruct
+	private void onStart()
+	{
+		int maxUsers = yonaProperties.getMaxUsers();
+		if (maxUsers <= 0)
+		{
+			return;
+		}
+
+		logger.info("Maximum number of users: {}", maxUsers);
+	}
+
+	@Transactional(dontRollbackOn = UserOverwriteConfirmationException.class)
+	public UserDto addUser(UserDto user, Optional<String> overwriteUserConfirmationCode)
+	{
+		assert user.getPrivateData().getDevices().orElse(Collections.emptySet()).size() == 1;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Replace this assert with a proper check. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://10.127.10.107:9000/coding_rules#rule_key=squid%3AS4274)